hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestBulkLoadReplication.java

```diff
@@ -0,0 +1,304 @@
+    ReplicationPeerConfig peer3Config = getPeerConfigForCluster(UTIL3);
+    //adds cluster1 as a remote peer on cluster2
+    UTIL2.getAdmin().addReplicationPeer(PEER_ID1, peer1Config);
+    //adds cluster3 as a remote peer on cluster2
+    UTIL2.getAdmin().addReplicationPeer(PEER_ID3, peer3Config);
+    //adds cluster2 as a remote peer on cluster3
+    UTIL3.getAdmin().addReplicationPeer(PEER_ID2, peer2Config);
+    setupCoprocessor(UTIL1);
+    setupCoprocessor(UTIL2);
+    setupCoprocessor(UTIL3);
+  }
+
+  private ReplicationPeerConfig getPeerConfigForCluster(HBaseTestingUtility util) {
+    return ReplicationPeerConfig.newBuilder()
+      .setClusterKey(util.getClusterKey()).setSerial(isSerialPeer()).build();
+  }
+
+  private void setupCoprocessor(HBaseTestingUtility cluster){
+    cluster.getHBaseCluster().getRegions(tableName).forEach(r -> {
+      try {
+        r.getCoprocessorHost()
+          .load(TestBulkLoadReplication.BulkReplicationTestObserver.class, 0,
+            cluster.getConfiguration());
+      } catch (Exception e){
+        LOG.error(e.getMessage(), e);
+      }
+    });
+  }
+
+  @After
+  @Override
+  public void tearDownBase() throws Exception {
+    super.tearDownBase();
+    UTIL2.getAdmin().removeReplicationPeer(PEER_ID1);
+    UTIL2.getAdmin().removeReplicationPeer(PEER_ID3);
+    UTIL3.getAdmin().removeReplicationPeer(PEER_ID2);
+  }
+
+  private static void setupBulkLoadConfigsForCluster(Configuration config,
+    String clusterReplicationId) throws Exception {
+    config.setBoolean(HConstants.REPLICATION_BULKLOAD_ENABLE_KEY, true);
+    config.set(REPLICATION_CLUSTER_ID, clusterReplicationId);
+    File sourceConfigFolder = testFolder.newFolder(clusterReplicationId);
+    File sourceConfigFile = new File(sourceConfigFolder.getAbsolutePath()
+      + "/hbase-site.xml");
+    config.writeXml(new FileOutputStream(sourceConfigFile));
+    config.set(REPLICATION_CONF_DIR, testFolder.getRoot().getAbsolutePath());
+  }
+
+  @Test
+  public void testBulkLoadReplicationActiveActive() throws Exception {
+    Table peer1TestTable = UTIL1.getConnection().getTable(TestReplicationBase.tableName);
+    Table peer2TestTable = UTIL2.getConnection().getTable(TestReplicationBase.tableName);
+    Table peer3TestTable = UTIL3.getConnection().getTable(TestReplicationBase.tableName);
+    byte[] row = Bytes.toBytes("001");
+    byte[] value = Bytes.toBytes("v1");
+    assertBulkLoadConditions(row, value, UTIL1, peer1TestTable, peer2TestTable, peer3TestTable);
+    row = Bytes.toBytes("002");
+    value = Bytes.toBytes("v2");
+    assertBulkLoadConditions(row, value, UTIL2, peer1TestTable, peer2TestTable, peer3TestTable);
+    row = Bytes.toBytes("003");
+    value = Bytes.toBytes("v3");
+    assertBulkLoadConditions(row, value, UTIL3, peer1TestTable, peer2TestTable, peer3TestTable);
+    //Additional wait to make sure no extra bulk load happens
+    Thread.sleep(400);
```

Comment:
Any way we could know when HBase is done replicating? Could we figure this out from the replication queue?

Reply:
Ick, yeah, that's hard to capture. I like the idea of proactively failing when we get more bulk loads than we expect (we can probably rule out retriable bulk load errors in a contrived unit test like this). Don't consider this to be some blocking request for change -- just an open question :)

Reply:
> (we can probably rule out retriable bulk load errors in a contrived unit test like this).

Moved the  BULK_LOADS_COUNT increment to _postBulkLoadHFile_ Observer method. That way, we would only be counting succeeded bulkloads, ignoring failed ones, so if it's a retry, the failed attempt didn't get count. 

But the problem of how to check something that shouldn't happen anytime in the future without the wait/sleep seems unsolvable to me.

Reply:
hehe, yeah. The only thing I could think of would be instrumenting the code that "skips" replicating to a cluster which already got it, but that's pretty down in the weeds, I'd say.

I think you're good :)